### PR TITLE
Add FSNER example in research_projects

### PR DESCRIPTION
**Pull request by sayef**
# What does this PR do?

- This PR adds example code for FSNER (few-shot named entity recognition) using huggingface's `transformers` library.
- Only prediction/inference code is provided, training code will be provided very soon.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
https://github.com/huggingface/transformers/pull/13133
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@NielsRogge @LysandreJik
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/research_projects/fsner/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+> We present a novel approach to named entity recognition (NER) in the presence of scarce data that we call example-based NER. Our train-free few-shot learning approach takes inspiration from question-answering to identify entity spans in a new and unseen domain. In comparison with the current state-of-the-art, the proposed method performs significantly better, especially when using a low number of support examples.
+
+
+
+## Model Training Details
+-----
+
+| identifier        | epochs           | datasets  |
+| ---------- |:----------:| :-----:|
+| sayef/fsner-bert-base-uncased      | 10 | ontonotes5, conll2003, wnut2017, and fin (Alvarado et al.). |
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:22]: Perhaps you can make sayef/fsner-bert-base-uncased a hyperlink to the model on the hub

File: examples/research_projects/fsner/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+
+## Installation and Example Usage
+------
+
+You can use the FSNER model in two ways-
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:28]:
```suggestion
You can use the FSNER model in two ways:
```

File: examples/research_projects/fsner/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+1. Install as a package: `python setup.py install` and import as following code example -
+
+    or
+
+2. Change directory to `src` and import as following code example -
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:34]:
```suggestion
2. Change directory to `src` and import the model as shown in the code example below
```

File: examples/research_projects/fsner/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+| ---------- |:----------:| :-----:|
+| sayef/fsner-bert-base-uncased      | 10 | ontonotes5, conll2003, wnut2017, and fin (Alvarado et al.). |
+
+
+## Installation and Example Usage
+------
+
+You can use the FSNER model in two ways-
+
+1. Install as a package: `python setup.py install` and import as following code example -
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:30]:
```suggestion
1. Install as a package: `python setup.py install` and import the model as shown in the code example below
```

File: examples/research_projects/fsner/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+
+```python
+from fsner import FSNERModel, FSNERTokenizerWrapper
+
+model = FSNERModel("sayef/fsner-bert-base-uncased")
+
+tokenizer = FSNERTokenizerUtils("sayef/fsner-bert-base-uncased")
+
+# size of query and supports must be same. If you want to find all the entitites in one particular query, just repeat same query n times where n is the size of supports (or entities).
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:45]:
```suggestion
# size of query and supports must be the same. If you want to find all the entitites in one particular query, just repeat the same query n times where n is equal to the number of supports (or entities).
```

File: examples/research_projects/fsner/setup.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import setuptools
+
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/setup.py:3]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/model.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import torch
+
+from transformers import AutoModel, AutoTokenizer
+
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:5]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/tokenizer_utils.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+from transformers import AutoTokenizer
+
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/tokenizer_utils.py:3]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/model.py (hunk 7 above)
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:5]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/model.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        return self.bert(**inputs).last_hidden_state
+
+    def VectorSum(self, token_embeddings):
+        return token_embeddings.sum(2, keepdim=True)
+
+    def Atten(self, q_rep, S_rep, T=1):
+        return self.softmax(T * self.cos(q_rep, S_rep))
+
+    def forward(self, W_query, W_supports):
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:30]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/model.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        """
+        Find scores of each token being start and end token for an entity.
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:32]:
```suggestion
        Find scores of each token being the start and end token for an entity.
```

File: examples/research_projects/fsner/src/fsner/model.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                Indices of query sequence tokens in the vocabulary.
+            W_supports (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+                Indices of support sequence tokens in the vocabulary.
+        Returns:
+            p_start (`torch.FloatTensor` of shape `(batch_size, sequence_length)`): Scores of each token as
+            being start token of an entity
+            p_end (`torch.FloatTensor` of shape `(batch_size, sequence_length)`): Scores of each token as
+            being end token of an entity
+        """
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:44]:
```suggestion
```

File: examples/research_projects/fsner/src/fsner/model.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        q = q.view(q.shape[0], -1, q.shape[1], q.shape[2])
+
+        # reshape from (batch_size*n_exaples_per_entity, 384, 784) to (batch_size, n_exaples_per_entity, 384, 784)
+        S = S.view(q.shape[0], -1, S.shape[1], S.shape[2])
+
+        s_start = S[(W_supports["input_ids"] == 30522).view(S.shape[:3])].view(S.shape[0], -1, 1, S.shape[-1])
+        s_end = S[(W_supports["input_ids"] == 30523).view(S.shape[:3])].view(S.shape[0], -1, 1, S.shape[-1])
+
+        p_start = torch.sum(torch.einsum("bitf,bejf->bet", q, s_start), dim=1)
+        p_end = torch.sum(torch.einsum("bitf,bejf->bet", q, s_end), dim=1)
```
Comment (NielsRogge) [inline examples/research_projects/fsner/src/fsner/model.py:58]: Can you also explain what is happening here?

**Comment by NielsRogge** [comment]
Looks great already! I left some small comments.

File: examples/research_projects/fsner/README.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        [
+            "[E] Walmart [/E] is a leading e-commerce company",
+            "I recently ordered a book from [E] Amazon [/E]",
+            "I ordered this from [E] ShopClues [/E]",
+            "Fridge can be ordered in [E] Amazon [/E]",
+            "[E] Flipkart [/E] started it's journey from zero"
+        ]
+   ]
+
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:75]:
```suggestion
```

File: examples/research_projects/fsner/README.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+query = [
+    'KWE 4000 can reach with a maximum speed from up to 450 P/min an accuracy from 50 mg',
+    'I would like to order a computer from eBay.',
+]
+
+# each list in supports are the examples of one entity type
+
+# wrap entities around with [E] and [/E] in the examples
+
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:57]:
```suggestion
```

File: examples/research_projects/fsner/README.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+# size of query and supports must be same. If you want to find all the entitites in one particular query, just repeat same query n times where n is the size of supports (or entities).
+
+
+query = [
+    'KWE 4000 can reach with a maximum speed from up to 450 P/min an accuracy from 50 mg',
+    'I would like to order a computer from eBay.',
+]
+
+# each list in supports are the examples of one entity type
+
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:54]:
```suggestion
```

File: examples/research_projects/fsner/README.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+2. Change directory to `src` and import as following code example -
+
+
+
+```python
+from fsner import FSNERModel, FSNERTokenizerWrapper
+
+model = FSNERModel("sayef/fsner-bert-base-uncased")
+
+tokenizer = FSNERTokenizerUtils("sayef/fsner-bert-base-uncased")
```
Comment (NielsRogge) [inline examples/research_projects/fsner/README.md:43]: I see you're importing `FSNERTokenizerWrapper`, but then using `FSNERTokenizerUtils`.

File: examples/research_projects/fsner/src/fsner/model.py (hunk 12 above)
Comment (sayef) [reply examples/research_projects/fsner/src/fsner/model.py:55]: Sure!

1. First, we get the query and support embeddings:

```
q = self.BERT(**W_query)
S = self.BERT(**W_supports)
```

2. Next, query and supports are reshaped for further easiness of calculations. The size of query becomes `(batch_size, 1, num_tokens, embedding_size)` and the size of supports becomes `(batch_size, n_examples_per_query, num_tokens, embedding_size)`

```
# reshape from (batch_size, 384, 784) to (batch_size, 1, 384, 784)
q = q.view(q.shape[0], -1, q.shape[1], q.shape[2])

# reshape from (batch_size*n_exaples_per_entity, 384, 784) to (batch_size, n_exaples_per_entity, 384, 784)
S = S.view(q.shape[0], -1, S.shape[1], S.shape[2])
```

3. Now, for each of the support set, we take the embeddings of entity start token `([E])` and end token `([/E])` for all examples in the set.

```
s_start = S[(W_supports["input_ids"] == 30522).view(S.shape[:3])].view(S.shape[0], -1, 1, S.shape[-1])
s_end = S[(W_supports["input_ids"] == 30523).view(S.shape[:3])].view(S.shape[0], -1, 1, S.shape[-1])
```

4. Lastly, find the probability of each token in the query being the start/end token by applying the following formula from the paper. I used `einsum` because I like it. Moreover, it helped me to multiply with soft attention, which is needed in training time. Also, the unused methods `VectorSum` and `Atten` is for training time. I will add training scripts soon.

image

```
p_start = torch.sum(torch.einsum("bitf,bejf->bet", q, s_start), dim=1)
p_end = torch.sum(torch.einsum("bitf,bejf->bet", q, s_end), dim=1)
```

Note 1: Simply to put, the proposed idea to find these probabilities is multiplying token level similarity with sentence level similarity as soft attention.

Note 2: The paper suggested a technique to be able to use a variable number of support examples, which I am currently not providing. I have plans to add that later too.

File: examples/research_projects/fsner/README.md (hunk 1 above)
Comment (sayef) [reply examples/research_projects/fsner/README.md]: I have added a hyperlink to the model on the hub.

File: examples/research_projects/fsner/README.md (hunk 16 above)
Comment (sayef) [reply examples/research_projects/fsner/README.md]: Mistake. Fixed in my latest [commit](https://github.com/huggingface/transformers/pull/13155/commits/bffa29558ed181a057a04db74f426f615219b4c5).

**Comment by sayef** [comment]
Hi @NielsRogge !

Would you mind telling me what else should I do? Or it's ready to merge?

Thanks!

File: examples/research_projects/fsner/src/fsner/model.py (hunk 12 above)
Comment (NielsRogge) [reply examples/research_projects/fsner/src/fsner/model.py:55]: Thanks for the explanation. However, I meant to perhaps add comments to the code to explain what it does. But now that you've explained it in detail, you can perhaps add this to the main README.

**Comment by NielsRogge** [comment]
Hi,

Now others need to review this, once they're back from holiday ;)

File: examples/research_projects/fsner/src/fsner/model.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+        q = self.BERT(**W_query)
+        S = self.BERT(**W_supports)
+
+        # reshape from (batch_size, 384, 784) to (batch_size, 1, 384, 784)
+        q = q.view(q.shape[0], -1, q.shape[1], q.shape[2])
+
+        # reshape from (batch_size*n_exaples_per_entity, 384, 784) to (batch_size, n_exaples_per_entity, 384, 784)
+        S = S.view(q.shape[0], -1, S.shape[1], S.shape[2])
+
+        s_start = S[(W_supports["input_ids"] == 30522).view(S.shape[:3])].view(S.shape[0], -1, 1, S.shape[-1])
```
Comment (stefan-it) [inline examples/research_projects/fsner/src/fsner/model.py:51]: Hi @sayef ,

this should be replaced with the "real" vocab size (instead of a magic number here), so that other BERT models with different vocab sized can be used :)

File: examples/research_projects/fsner/src/fsner/model.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+from transformers import AutoModel, AutoTokenizer
+
+class FSNERModel(torch.nn.Module):
+    """
+    The FSNER model implements a few-shot named entity recognition method from the paper `Example-Based Named Entity Recognition <https://arxiv.org/abs/2008.10570>`__ by
+    Morteza Ziyadi, Yuting Sun, Abhishek Goswami, Jade Huang, Weizhu Chen. To identify entity spans in a new domain, it
+    uses a train-free few-shot learning approach inspired by question-answering.
+    """
+
+    def __init__(self, pretrained_model_name_or_path="sayef/fsner-bert-base-uncsed"):
```
Comment (stefan-it) [inline examples/research_projects/fsner/src/fsner/model.py:12]:
```suggestion
    def __init__(self, pretrained_model_name_or_path="sayef/fsner-bert-base-uncased"):
```

File: examples/research_projects/fsner/src/fsner/model.py (hunk 17 above)
Comment (sayef) [reply examples/research_projects/fsner/src/fsner/model.py:51]: Thanks @stefan-it, I also thought about it. 

These two numbers are the special token IDs for [E] and [/E] which was added in the training time. 

Now, one way to do that: first, users have to get the ids using `tokenizer.tokens_to_ids(['[E]', '[/E]'])` and then pass it to forward function. 

The other way is to use model config's `vocab_size` as this number and `vocalb_size-1` would be those token ids, since the last added extra tokens are assigned incrementally. 

Which would you suggest? First one seems elegant, but users have to do it by themselves. The latter one seems hackish, but users don't have to worry about it.

**Review by LysandreJik — APPROVED** [review]
This is great, thanks a lot @sayef!

**Comment by LysandreJik** [comment]
Hi @sayef, there are a few code quality issues; I tried to push to your branch but I do not have push access on your branch.

Could you run the following commands at the root of your clone? It should tell you what needs fixing:
```
pip install -U -e .[quality]
make fixup
```

**Comment by sayef** [comment]
Hi @LysandreJik,

I followed what you suggested. Let me know if I need to do anything else. :)

**Comment by LysandreJik** [comment]
Hi @sayef  - I believe you also rebased or merged the `master` branch into your PR. Unfortunately, GitHub sometimes has issues understanding what happened, for example here your PR shows 245 commits and 466 files changed. 

Usually just closing the PR and opening a new one from the same branch, without changing anything is enough. Would you mind doing that and pinging me so that I may merge? Thank you!

**Comment by sayef** [comment]
> Usually just closing the PR and opening a new one from the same branch, without changing anything is enough. Would you mind doing that and pinging me so that I may merge? Thank you!

Okay. Closing here and will ping you in other PR.
